Update ghostdog to handle multiple neuron drivers

yeazelm opened the pull request:
**Issue number:**

Related to https://github.com/bottlerocket-os/bottlerocket-kernel-kit/issues/238

**Description of changes:**
This adds a `match-driver` command to ghostdog which will eventually replace `match-nvidia-driver` as a generic command for multiple drivers. This allows neuron devices to be chosen in the same way that NVIDIA drivers are based upon the devices found on the PCI bus.

This is useful for inf1 instances where they will need to remain on 2.21.* but newer instances would benefit from a newer driver. This will allow Bottlerocket to choose the best driver for the particular instance it is launched on. This change relies upon a change in the kernel kit to vend this additional driver and call `driverdog` to actually do the loading.

See https://github.com/bottlerocket-os/bottlerocket-kernel-kit/pull/277 for the other side of this change. 


**Testing done:**
Booted an image with this and https://github.com/bottlerocket-os/bottlerocket-kernel-kit/pull/277 and booted on both inf1 and trn1 instances. The correct driver is loaded on each.

### on `trn1.2xlarge'
```
[ssm-user@control]$ apiclient exec admin modinfo neuron
filename:       /lib/modules/6.12.40/neuron_latest/neuron.ko
alias:          pci:v00001d0fd00007064sv*sd*bc*sc*i*
version:        2.23.9.0
license:        GPL
description:    Neuron Driver, built from SHA: b8ad53947481bbe9d8e76b117c06b2484b8a949f
import_ns:      DMA_BUF
srcversion:     68886FE4A17A11B0934DAFE
depends:
name:           neuron
retpoline:      Y
vermagic:       6.12.40 SMP preempt mod_unload modversions
```

### on `inf1.xlarge`
```
[ssm-user@control]$ apiclient exec admin modinfo neuron
filename:       /lib/modules/6.12.40/neuron_2_21/neuron.ko
alias:          pci:v00001d0fd00007064sv*sd*bc*sc*i*
version:        2.21.37.0
license:        GPL
description:    Neuron Driver, built from SHA: fd7535995e4b3ae73c543f829b152c3351ef21f6
import_ns:      DMA_BUF
srcversion:     4DE159506B58B4692C035AA
depends:
name:           neuron
retpoline:      Y
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
pciclient: add inf1 instance functions

Add functions specifically for inf1 instances to allow detecting the
difference between any Neuron device and inf1 devices.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

yeazelm commented on sources/driverdog/src/main.rs:
```diff
@@ -403,6 +432,43 @@ fn get_kernel_version() -> Result<String> {
+        Ok(OpinionatedModule::Neuron) => {
+            // get data from pciclient for inf1
+            if is_inf1_instance().context(error::CheckInf1FailureSnafu)? {
+                if !Path::new(DEPMOD_D_PATH).is_dir() {
```
We should switch to tmpfiles directive.

bcressey commented on sources/pciclient/src/private.rs:
```diff
@@ -27,6 +27,15 @@ lazy_static! {
 }
 
+lazy_static! {
+    static ref INF1_DEVICES: HashSet<&'static str> = hashset! {
+        "7064", // inf1 device id 0
+        "7065", // inf1 device id 1
+        "7066", // inf1 device id 2
+        "7067", // inf1 device id 3
+    };
+}
```
Rather than duplicating this, I'd rather split the current `NEURON_DEVICES` into two sets, and update the `check_neuron_attachment` function to query both.

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -69,6 +74,7 @@ struct Args {
 enum Subcommand {
     Link(LinkModulesArgs),
     Load(LoadModulesArgs),
+    Prioritize(PrioritizeModuleArgs),
```
I'd rather use a `ghostdog` subcommand helper for this, and if possible do it the same way we do NVIDIA driver matching: systemd units with a `ghostdog` ExecCondition that invoke `driverdog` with different configs.

Commit:
release: Add drivers.target

Add a target for loading drivers so we can isolate dependencies strictly
to when all drivers are loaded separately from preconfigured.target.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
ghostdog: Add match-driver command

This command is designed to be a more generic version of
match-nvidia-driver and will eventually replace this functionality once
all consumers are using the new command. This command also supports
neuron devices and homogenizes the approach for choosing drivers based
upon PCI devices.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

bcressey commented on packages/release/drivers.target:
```diff
@@ -0,0 +1,8 @@
+After=basic.target
+Requires=basic.target
+
+[Install]
+RequiredBy=preconfigured.target multi-user.target
```
This doesn't establish an ordering dependency with `preconfigured.target`, and we want it to happen before.

yeazelm replied:
Yeah this worked in my tests but you are right that we need `Before=preconfigured.target` to ensure it really does go before.

bcressey commented on sources/pciclient/src/lib.rs:
```diff
@@ -147,4 +152,6 @@ mod error {
 
 pub use error::PciClientError;
+
+use crate::private::check_inf1_attachment;
```
This should go up at the top with the other, non-error related `use` statements.

yeazelm replied:
I'll fix, not sure why the auto add from my IDE chose here.

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -359,6 +404,17 @@ fn nvidia_driver_supported(driver_name: &str) -> Result<()> {
+    if driver_flavor == "inf1" {
+        return is_inf1_instance();
+    }
+    if driver_flavor == "latest" {
+        return is_after_inf1_instance();
+    }
```
Let's also return an error on an unmatched flavor. If we don't return non-zero then the `ExecCondition` logic in units won't work as expected.

yeazelm replied:
Make sense!

bcressey commented on sources/Cargo.lock:
```diff
@@ -2317,7 +2318,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 dependencies = [
  "libc",
- "windows-sys 0.52.0",
+ "windows-sys 0.59.0",
```
Can we back out the `windows-sys` changes?

yeazelm replied:
Yes, I'll unwind this change.

bcressey commented on sources/driverdog/Cargo.toml:
```diff
@@ -11,6 +11,7 @@ exclude = ["README.md"]
 [dependencies]
 argh.workspace = true
 log.workspace = true
+pciclient.workspace = true
```
This change is no longer needed?

yeazelm replied:
Correct, I'll clean up the Cargo dependencies from the refactor.

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -138,13 +153,19 @@ fn link_modules_sets(
         match driver_config {
             DriverType::Copying(config) => copy_modules(config, &kernel_version)?,
             DriverType::Linking(config) => link_modules(config, &kernel_version)?,
+            DriverType::DepmodOnly(_) => {
+                debug!("Depmod only is not supported for linking modules sets")
+            }
```
Can we call `configure_modules` here so that `DepmodOnly` can be handled the same way as the other two? And then `load_modules` would have the same boilerplate to extract the list of modules to load.

Conceptually `driverdog` has two stages: "prepare" (which we call "link") and "load", so I'd prefer to do the work in the "prepare" stage rather than deferring it to "load".

Or in other words: the only side effect of `driverdog load-modules` should be that modules might be loaded. Other filesystem operations should happen in `driverdog link-modules`.

yeazelm replied:
"link" is really doing too much here now, but I agree that had we named it "prepare", this would feel natural. In this case I do agree we should prep the depmod in "link" and only load in "load". I debated on if we wanted to have two invocations of driverdog for these but I think the additional call is worth it to keep the separation of concerns. I'll update the logic to do so.

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -403,6 +428,65 @@ fn get_kernel_version() -> Result<String> {
+    let config_string = module
+        .kernel_modules
+        .iter()
+        .map(|name| format!("override {} * {}", name, module.lib_modules_path))
```
Can we also set the specific kernel version instead of `*` here, since we know what that version is?

yeazelm replied:
Yeah, we can be more specific. We don't ship more than one kernel in an image so it isn't strictly required but this does make it more correct and specific.

yeazelm replied:
I added this logic, the resulting file at `/etc/depmod.d/neuron.conf` now looks like this for inf1:
```
override neuron 6.12.40 neuron_2_21
```

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -438,20 +522,22 @@ fn run() -> Result<()> {
-            &fs::read_to_string(&path).context(error::ReadPathSnafu { path: &path })?,
-        )
-        .context(error::DeserializeSnafu { path: &path })?;
+    if Path::new(driver_config_path).is_dir() {
+        for entry in (driver_config_path
```
Why do we need this new conditional logic? Why wouldn't `driver_config_path` exist?

yeazelm replied:
This change was a quick one I made when I first added driverdog to all images. We didn't provide `/etc/drivers` since that was only created by NVIDIA kmods. In this case, I have refactored since then to provide files for neuron so we don't strictly need this check anymore, but it does feel overly strict to not allow driverdog to run *at all* without `/etc/drivers` existing on the system. Since `driver_config_path` is always created, driverdog _always_ tries to load these files. I think a more correct way to solve this is to package a tmpfiles directive that creates this directory with driverdog since the tool has some deep assumptions about these configuration files being present to fully function. There isn't really a valid use case to use driverdog without these files being present. I can back it out, or we can ensure the directory is present as part of packaging.

bcressey replied:
> the tool has some deep assumptions about these configuration files being present to fully function

I'd just do an early return with a "no configurations found in `/etc/drivers`" error message, if the path doesn't exist.

yeazelm commented:
^ Updated the code to reflect comments. There is now two calls, one for link and one for load in driverdog. The kernel kit PR has been updated as well to include these calls. I re-ran validation that the expected drivers load on each instance type.

bcressey commented on packages/release/drivers.target:
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=Bottlerocket driver loading complete
```
From existing target units, maybe just this:
```suggestion
Description=Driver units
```

We don't want a message like "Started Bottlerocket driver loading complete" on the console.

bcressey commented on sources/ghostdog/src/error.rs:
```diff
@@ -66,6 +66,16 @@ pub(super) enum Error {
         requested: String,
         preferred: String,
     },
+    #[snafu(display("{driver} is not a supported driver"))]
+    UnsupportedDriver { driver: String },
+    #[snafu(display("{flavor} is not a supported choice for {driver}"))]
```
Do these `{driver}` and `{source}` format parameters work?

I don't recall seeing them elsewhere.

yeazelm replied:
Clippy has recently been calling out the other style of `#[snafu(display("{} is not a supported choice for {}", flavor, driver))]` in explicit `format!()` calls, I just started getting into a pattern of using this and sure enough it worked without me thinking about it. I can try moving it back but I actually do like this style. We can move everything to it in a follow up commit/PR if we want to keep things consistent. 

I wanted to double check and sure enough, it works as expected:
```
bash-5.1# /usr/bin/ghostdog match-driver foo bar
Error: foo is not a supported driver

bash-5.1# /usr/bin/ghostdog match-driver neuron bar
Error: bar is not a supported choice for neuron

bash-5.1#
```

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -179,6 +203,27 @@ fn is_neuron_attached() -> Result<()> {
+        }
+    }
+    Err(error::Error::NoNeuronPresent)
+}
```
I see why you're doing this, to make the error messages more friendly, but it's going to be hard to maintain that if we ever need a third driver variant.

I would probably just have `is_inf1_instance` call `pciclient::is_inf1_instance()`, and `is_inf1_instance` call `pciclient::is_inf2_instance()`, and let "inf2" be a stand-in for all hardware of that generation until the next breaking change. (IIRC it was like this earlier.) Or `trn1` instead of `inf2`.

As it stands, the function ends up calling `lspci` twice, once with a combined Neuron devices filter, and again with a Inf1-only filter, and that's suboptimal.

yeazelm replied:
I agree. This may yield a bit less data but really streamlines the implementation. I'll shift everything to be specific to inf2. This name doesn't cover everything as of today but leaves it specific enough that any future refactors should be easier to make.

bcressey commented on sources/pciclient/src/private.rs:
```diff
@@ -166,13 +173,27 @@ pub(crate) fn check_efa_attachment<T: CommandExecutor>(command_executor: T) -> R
+    }))
+}
+
+/// Call `lspci` and check if there is any inf1-specific Neuron device attached.
+/// Internal usage, adding command_executor as parameter allows us to better unit test.
+/// For external usage, check [`crate::is_neuron_attached`].
+pub(crate) fn check_inf1_attachment<T: CommandExecutor>(command_executor: T) -> Result<bool> {
```
The `check_neuron_attachment` change is good for backwards compatibility; `ghostdog neuron-present` should probably continue to work for now. But I'd still add a `check_inf2_attachment` function and related helper, and have `ghostdog` call `is_inf2_attached`, even if it means a third near-duplication of this method.

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -138,13 +153,17 @@ fn link_modules_sets(
                 DriverType::Linking(config) => link_modules(config, &kernel_version)?,
+                DriverType::DepmodOnly(_) => {
+                    debug!("Depmod only is not supported for linking modules sets")
+                }
```
I'd prefer to do the depmod configuration at "link" time, and keep the load path consistent across the three driver types.

yeazelm replied:
I'll update this. I was overly focused on the single case and didn't shift this to be consistent.

arnaldo2792 commented on sources/ghostdog/src/error.rs:
```diff
@@ -66,6 +66,18 @@ pub(super) enum Error {
         requested: String,
         preferred: String,
```
nit: typo in commit message:

```
This command also supports
```

arnaldo2792 commented on sources/ghostdog/src/main.rs:
```diff
@@ -156,6 +167,19 @@ fn main() -> Result<()> {
+                "nvidia" => nvidia_driver_supported(&flavor_name)?,
+                "neuron" => match_neuron_driver(&flavor_name)?,
+                _ => {
+                    return Err(error::Error::UnsupportedDriver {
+                        driver: driver_name.to_string(),
+                    })
```
I think you can do:

```suggestion
                    return error::Error::UnsupportedDriver {
                        driver: driver_name.to_string(),
                    }.fail()
```

arnaldo2792 commented on sources/ghostdog/src/main.rs:
```diff
@@ -359,6 +402,20 @@ fn nvidia_driver_supported(driver_name: &str) -> Result<()> {
+    Err(error::Error::UnsupportedDriverFlavor {
+        driver: "neuron".to_string(),
+        flavor: driver_flavor.to_string(),
+    })
```
Same as elsewhere, you could just `fail`.

yeazelm replied:
The rest of ghostdog is using `Err()` style. I can shift to fail() but we should remain consistent in the program. If we like `fail()` more than this style, I can do a change to follow up with the style improvement.

arnaldo2792 commented on sources/ghostdog/src/main.rs:
```diff
@@ -359,6 +402,20 @@ fn nvidia_driver_supported(driver_name: &str) -> Result<()> {
 }
 
+fn match_neuron_driver(driver_flavor: &str) -> Result<()> {
+    if driver_flavor == "inf1" {
```
Why not use `match` for `inf1` and `latest` and fail otherwise in the fallback branch?

yeazelm replied:
This was to keep it _mostly_ consistent with the rest of the if statements in these functions. I think an enum for supported flavors is probably a stronger mechanism anyway and then the match statement can work from that. It will force the compiler to hold us accountable when adding any new flavors. I'll update with this change!

yeazelm replied:
The flavor changes depending on the original driver. We may want to refactor this to be more robust in types for these flavors but for now its two strings, so I'll keep the strings but in the future if we add more, we should define types to hold all of these.

arnaldo2792 commented on sources/ghostdog/src/main.rs:
```diff
@@ -179,6 +203,25 @@ fn is_neuron_attached() -> Result<()> {
 
+// Returns true if this is an inf1 instance
+fn is_inf1_instance() -> Result<()> {
+    if pciclient::is_inf1_instance().context(error::CheckInf1FailureSnafu)? {
```
Wouldn't this error be more "correct" in the underlying `check_` function? then in here you can do `pciclient::is_inf1_instance()?` without having to add any context.

yeazelm replied:
This is following the existing style of other functions like `is_efa_attached()` or `is_neuron_attached()`. The intention here is what function failed *is* the context. The failure would ultimately be coming from being unable to use `lspci` and so this helps provide the additional context on what check failed along with the actual failure. So the `check_` functions will emit logs about the `call_list_devices()` call failing and the additional messaging will be why we were calling `call_list_devices()`. I think this might be a bigger refactor of error handling in pciclient which we can do if we think it will improve error messaging, but probably should do it separately from this change.

arnaldo2792 commented on sources/ghostdog/src/main.rs:
```diff
@@ -179,6 +203,25 @@ fn is_neuron_attached() -> Result<()> {
+// inf1 but leaves this open for more hardware types that may need specific decisions from ghostdog in the
+// future.
+fn is_inf2_instance() -> Result<()> {
+    if pciclient::is_inf2_instance().context(error::CheckInf2FailureSnafu)? {
+        return Ok(());
+    }
+    Err(error::Error::NoInf2Present)
```
Above you implemented this with `if/else`. For consistency, you should use the same here.

bcressey replied:
Agreed re: consistency but I prefer this style - early return + outdented body for what would be the else clause. So I'd make the other one look like this rather than vice versa.

bcressey reviewed (approved):
LGTM, no notes beyond what @arnaldo2792 has commented on.
